### PR TITLE
feat: add component filter

### PR DESCRIPTION
**Pull request by 0xff-dev**
Fix: https://github.com/kubebb/core/issues/19

### CODE REVIEW

File: api/v1alpha1/common_test.go [hunk 1]
```diff
@@ -312,3 +312,101 @@ func TestUpdateCondwithFixedLen(t *testing.T) {
...
+		{
+			description: "cond.Deprecated为false且deprecated为true",
+			name:        "test",
+			filterCond:  map[string]FilterCond{"test": {Deprecated: false}},
+			version:     "1.0.0",
+			deprecated:  true,
+			expected:    false,
+		},
+		{
+			description: "version存在于cond.Versions",
```
Comment (bjwswang) [inline api/v1alpha1/common_test.go:350]: CN -> English

File: api/v1alpha1/common.go [hunk 2]
```diff
@@ -72,3 +77,48 @@ func UpdateCondWithFixedLen(l int, conds *ConditionedStatus, cond Condition) {
 	}
 	conds.Conditions = append(conds.Conditions, cond)
 }
+
+func Match(filterCond map[string]FilterCond, name, version string, deprecated bool) bool {
```
[the remark this answers is not in the document]
Comment (bjwswang) [reply api/v1alpha1/common.go:81]: This `Match` is good as well!We can leave this until it really needs to be optimized.

**Review by bjwswang — APPROVED** [review]

File: api/v1alpha1/common.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (bjwswang) [reply api/v1alpha1/common.go:81]: Does this `c.Check` supports ">=1.2.3 & <=2.1.3" at the same time?

File: api/v1alpha1/repository_types.go [hunk 3]
```diff
@@ -32,6 +32,19 @@ type PullStategy struct {
...
+type FilterCond struct {
+	// If True, the current version will be retained even if it is deprecated.
+	Deprecated bool `json:"deprecated,omitempty"`
+	// Accurately match each item in the versions
+	Versions []string `json:"versions,omitempty"`
+	// Filter version by regexp
+	Regexp string `json:"regexp,omitempty"`
+
+	// All versions smaller/greater than a particular version
+	Than string `json:"than,omitempty"`
```
Comment (bjwswang) [inline api/v1alpha1/repository_types.go:44]: add a comment to point this constraint to `https://github.com/Masterminds/semver#semver`

**Review by bjwswang — APPROVED** [review]

File: api/v1alpha1/common.go [hunk 4]
```diff
@@ -72,3 +77,109 @@ func UpdateCondWithFixedLen(l int, conds *ConditionedStatus, cond Condition) {
...
+	if filterMatchVersionConstraint(cond, filter) {
+		return true
+	}
+	return false
+}
+
+func IsCondSame(c1, c2 FilterCond) bool {
+	c1Map := make(map[string]int)
+	c2Map := make(map[string]int)
+	for _, v := range c1.Versions {
```
Comment (0xff-dev) [inline api/v1alpha1/common.go:151]: @bjwswang @Abirdcfly Is it reasonable to implement this to determine if the contents of an array are consistent?

Comment (bjwswang) [reply api/v1alpha1/common.go:151]:
```golang
func stringSlicesEqual(s1, s2 []string) bool {
    if len(s1) != len(s2) {
        return false
    }
    set1 := make(map[string]bool)
    for _, v := range s1 {
        set1[v] = true
    }
    for _, v := range s2 {
        if !set1[v] {
            return false
        }
    }
    return true
}

```

Comment (Abirdcfly) [reply api/v1alpha1/common.go:151]: just for versions compare.
```go
// "k8s.io/apimachinery/pkg/util/sets"
sets.NewString(c1.Versions).Equal(sets.NewString(c2.Versions))
```

Comment (bjwswang) [reply api/v1alpha1/common.go:151]: /wtd verify two string slice equals or not

File: api/v1alpha1/common.go [hunk 5]
```diff
@@ -72,3 +77,109 @@ func UpdateCondWithFixedLen(l int, conds *ConditionedStatus, cond Condition) {
...
+	}
+	if filterMatchVersionConstraint(cond, filter) {
+		return true
+	}
+	return false
+}
+
+func IsCondSame(c1, c2 FilterCond) bool {
+	c1Map := make(map[string]int)
+	c2Map := make(map[string]int)
```
Comment (bjwswang) [inline api/v1alpha1/common.go:150]: /wtd verify two string slice equals or not

Comment (bjwswang) [reply api/v1alpha1/common.go:150]: /wtd compare two FilterCond

**Comment by bjwswang** [comment]
:wtd summary

File: api/v1alpha1/common.go (hunk 4 above)
Comment (0xff-dev) [reply api/v1alpha1/common.go:151]: > ```go
> func stringSlicesEqual(s1, s2 []string) bool {
>     if len(s1) != len(s2) {
>         return false
>     }
>     set1 := make(map[string]bool)
>     for _, v := range s1 {
>         set1[v] = true
>     }
>     for _, v := range s2 {
>         if !set1[v] {
>             return false
>         }
>     }
>     return true
> }
> ```

version1 = [v1, v1, v2]
version2 = [v2, v2, v2]
should return false, but.

**Review by Abirdcfly — APPROVED** [review]

**Pull request merged**